detect/gcp/gcp.go

```diff
@@ -0,0 +1,90 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"context"
+	"log"
+	"strings"
+
+	"cloud.google.com/go/compute/metadata"
+
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/api/standard"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+// GCP is a detector that implements Detect() function
+type GCP struct{}
+
+// Detect detects associated resources when running on GCE hosts.
+func (gcp *GCP) Detect(ctx context.Context) (*resource.Resource, error) {
+	if !metadata.OnGCE() {
+		return nil, nil
+	}
+
+	labels := []kv.KeyValue{
+		standard.CloudProviderGCP,
+		standard.CloudRegionKey.String(""),
+	}
+
+	projectID, err := metadata.ProjectID()
+	logError(err)
+	if projectID != "" {
+		labels = append(labels, standard.CloudAccountIDKey.String(projectID))
+	}
+
+	zone, err := metadata.Zone()
+	logError(err)
+	if zone != "" {
+		labels = append(labels, standard.CloudZoneKey.String(zone))
+	}
+
+	instanceID, err := metadata.InstanceID()
+	logError(err)
+	if instanceID != "" {
+		labels = append(labels, standard.HostIDKey.String(instanceID))
+	}
+
+	name, err := metadata.InstanceName()
+	logError(err)
+	if name != "" {
+		labels = append(labels, standard.HostNameKey.String(name))
+	}
+
+	hostname, err := metadata.Hostname()
```

Comment:
the spec at https://github.com/open-telemetry/opentelemetry-go/blob/d979977a406e01e2bb18dd510bdd12ebc4f4ffa7/api/standard/resource.go#L147 says that it's the result of running `hostname` on the machine, rather than a value pulled from the cloud provider.

what happens if I change hostname on the local machine post-boot without updating the cloud provider's metadata? this field should reflect the local machine's value as I understand.

Reply:
Good question. I tried to change host name on gce but the result was same. I doubted the definition. I don't think it collects host.hostname in that way. Here is the implementation. https://github.com/googleapis/google-cloud-go/blob/c07045b74f22a4ea094da4eb65e7b6413629df7e/compute/metadata/metadata.go#L386

Reply:
I think then that this package should either not set a host name resource value, deferring to another package to do that since it should be common across all providers, or it should use [`os.Hostname()`](https://golang.org/pkg/os/#Hostname).